pkg/controllers/workgenerator/reconciler.go

```diff
@@ -0,0 +1,299 @@
+	for _, work := range works {
+		if activeWork[work.Name] {
+			continue
+		}
+		klog.V(2).InfoS("delete the work that is not associated with any resource snapshot", "work", klog.KObj(work))
+		if err := r.Client.Delete(ctx, work); err != nil {
+			klog.ErrorS(err, "failed to delete the no longer needed work", "work", klog.KObj(work))
+			return ctrl.Result{}, controller.NewAPIServerError(false, err)
+		}
+	}
+
+	klog.V(2).InfoS("successfully synced all the work associated with the resourceBinding", "resourceBinding", klog.KObj(resourceBinding))
+	return ctrl.Result{}, nil
+}
+
+// fetchAllResourceSnapshots gathers all the resource snapshots for the resource binding
+func (r *Reconciler) fetchAllResourceSnapshots(ctx context.Context, resourceBinding *fleetv1beta1.ClusterResourceBinding) ([]*fleetv1beta1.ClusterResourceSnapshot, error) {
+	resourceSnapshots := make([]*fleetv1beta1.ClusterResourceSnapshot, 1)
+	resourceSnapshot := fleetv1beta1.ClusterResourceSnapshot{}
+	if err := r.Client.Get(ctx, client.ObjectKey{Name: resourceBinding.Spec.ResourceSnapshotName}, &resourceSnapshot); err != nil {
+		klog.ErrorS(err, "failed to get the resource snapshot from resource resourceBinding",
+			"resource resourceBinding", klog.KObj(resourceBinding), "resource snapshot name", resourceBinding.Spec.ResourceSnapshotName)
+		return nil, controller.NewAPIServerError(true, err)
+	}
+	resourceSnapshots[0] = &resourceSnapshot
+
+	// check if there are more snapshot in the same index group
+	countAnnotation := resourceSnapshot.Annotations[fleetv1beta1.NumberOfResourceSnapshotsAnnotation]
+	snapshotCount, err := strconv.Atoi(countAnnotation)
+	if err != nil {
+		return nil, controller.NewUnexpectedBehaviorError(err)
+	}
+	if snapshotCount > 1 {
+		index, exist := resourceSnapshot.Labels[fleetv1beta1.ResourceIndexLabel]
+		if !exist {
+			unexpectedErr := fmt.Errorf("resource snapshot %s has no index label", resourceSnapshot.Name)
+			return nil, controller.NewUnexpectedBehaviorError(unexpectedErr)
+		}
+		resourceIndexLabelMatcher := client.MatchingLabels{
+			fleetv1beta1.ResourceIndexLabel: index,
+		}
+		resourceSnapshotList := &fleetv1beta1.ClusterResourceSnapshotList{}
+		if err := r.Client.List(ctx, resourceSnapshotList, resourceIndexLabelMatcher); err != nil {
+			klog.ErrorS(err, "failed to list all the resource snapshot associated with the resourceBinding", "resourceBinding", klog.KObj(resourceBinding))
+			return nil, controller.NewAPIServerError(true, err)
+		}
+		for i := 0; i < snapshotCount; i++ {
+			resourceSnapshots = append(resourceSnapshots, &resourceSnapshotList.Items[i])
+		}
+	}
+	return resourceSnapshots, nil
+}
+
+// upsertWork creates or updates the work for the corresponding resource snapshot
+func (r *Reconciler) upsertWork(ctx context.Context, work *workv1alpha1.Work, resourceSnapshot *fleetv1beta1.ClusterResourceSnapshot, resourceBinding *fleetv1beta1.ClusterResourceBinding) error {
+	needCreate := false
```

Comment:
you could use the controllerutil.CreateOrUpdate to archive this in the 	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil" package

Reply:
I looked into this, the update function seems to be more complicated to write than this version

Reply:
hmm, i think it should be fine. we just need to ensure the labels, ownerreference and spec are set.
If i remember it correctly, member agent will only update the status of the work?